File: backend/src/api/news.ts
import { Hono } from 'hono';
import { randomUUID } from 'node:crypto';
import { listNews, affectedClientsForItem, seedNews, type NewsItem } from '../skills/news/index.js';

export const newsRoute = new Hono();

newsRoute.get('/', async (c) => {
  const items = listNews();
  const out = await Promise.all(
    items.map(async (it) => ({
      ...it,
      affected: await affectedClientsForItem(it),
    })),
  );
  return c.json({ items: out });
});

/**
 * Append (not replace) news items to the in-memory store.
 * Body: { items: NewsItem[] } — id + published_at are filled in if absent.
 */
newsRoute.post('/', async (c) => {
  const body = await c.req.json<{ items: Partial<NewsItem>[] }>().catch(() => ({ items: [] }));
  const incoming = (body.items ?? []).map<NewsItem>((it) => ({
    id: it.id ?? randomUUID(),
    title: it.title ?? '(untitled)',
    summary: it.summary ?? '',
    sectors: it.sectors ?? [],
    isins: it.isins ?? [],
    asset_class: (it.asset_class as NewsItem['asset_class']) ?? 'macro',
    published_at: it.published_at ?? new Date().toISOString(),
  }));
  seedNews([...listNews(), ...incoming]);
  return c.json({ ok: true, added: incoming.length, total: listNews().length });
});

File: backend_py/stackwealth/api/news.py
"""/api/news — list + append."""
from __future__ import annotations

from datetime import datetime, timezone
from uuid import uuid4

from fastapi import APIRouter, Request
from fastapi.responses import JSONResponse

from ..skills.news import affected_clients_for_item, list_news, seed_news

router = APIRouter()


@router.get("")
async def list_route() -> JSONResponse:
    items = list_news()
    out = []
    for it in items:
        affected = await affected_clients_for_item(it)
        out.append({**it, "affected": affected})
    return JSONResponse(content={"items": out})


@router.post("")
async def append(request: Request) -> JSONResponse:
    body = await request.json() if (await request.body()) else {"items": []}
    incoming = []
    for it in body.get("items") or []:
        incoming.append(
            {
                "id": it.get("id") or str(uuid4()),
                "title": it.get("title") or "(untitled)",
                "summary": it.get("summary") or "",
                "sectors": it.get("sectors") or [],
                "isins": it.get("isins") or [],
                "asset_class": it.get("asset_class") or "macro",
                "published_at": it.get("published_at")
                or datetime.now(timezone.utc).isoformat(),
            }
        )
    seed_news(list_news() + incoming)
    return JSONResponse(
        content={"ok": True, "added": len(incoming), "total": len(list_news())}
    )
